File: src/style/analyzer.ts
import type { RawMessage, StyleProfile, TypicalError } from "../corpus/types.js";

export function analyzeStyle(messages: RawMessage[]): StyleProfile {
  if (messages.length === 0) {
    throw new Error("No messages to analyze");
  }

  const texts = messages.map((m) => m.text);

  const avgLength =
    texts.reduce((sum, t) => sum + t.length, 0) / texts.length;

  const capsFirst =
    texts.filter((t) => /^[A-Z]/.test(t)).length / texts.length;

  const withContractions = texts.filter((t) =>
    /\b(i'm|i'll|i've|it's|that's|there's|don't|can't|won't|didn't|isn't|aren't|wasn't|weren't|hasn't|haven't|couldn't|wouldn't|shouldn't|y'all|let's|he's|she's|we're|they're|you're|who's|what's|where's|how's)\b/i.test(
      t
    )
  ).length;
  const contractionsRatio = withContractions / texts.length;

  const withPeriods =
    texts.filter((t) => /\.\s*$/.test(t)).length / texts.length;
  const withCommas = texts.filter((t) => t.includes(",")).length / texts.length;
  const withExclamation =
    texts.filter((t) => t.includes("!")).length / texts.length;
  const withQuestion =
    texts.filter((t) => t.includes("?")).length / texts.length;

  const emojiRegex =
    /[\u{1F600}-\u{1F64F}\u{1F300}-\u{1F5FF}\u{1F680}-\u{1F6FF}\u{1F1E0}-\u{1F1FF}\u{2600}-\u{26FF}\u{2700}-\u{27BF}]/u;
  const withEmoji =
    texts.filter((t) => emojiRegex.test(t)).length / texts.length;

  // sentence fragments: messages without a verb-like word and under 30 chars
  const fragmentish = texts.filter(
    (t) => t.length < 30 && t.split(/\s+/).length <= 5
  ).length;
  const fragmentRatio = fragmentish / texts.length;

  // lowercase "i" as a standalone word (not I'm, I'll, etc.)
  const lowercaseI = texts.filter((t) =>
    /\bi\b(?!')/.test(t)
  ).length;
  const iTotal = texts.filter((t) =>
    /\b[iI]\b/i.test(t)
  ).length;
  const lowercaseIRatio = iTotal > 0 ? lowercaseI / iTotal : 0;

  const commonPhrases = findCommonPhrases(texts);
  const typicalErrors = findTypicalErrors(texts);

  return {
    averageLength: Math.round(avgLength),
    capitalizesFirstWord: round(capsFirst),
    usesContractions: round(contractionsRatio),
    usesPeriods: round(withPeriods),
    usesCommas: round(withCommas),
    usesExclamation: round(withExclamation),
    usesQuestionMark: round(withQuestion),
    usesEmoji: round(withEmoji),
    commonPhrases,
    typicalErrors,
    sentenceFragmentRatio: round(fragmentRatio),
    lowercaseIRatio: round(lowercaseIRatio),
  };
}

function round(n: number): number {
  return Math.round(n * 1000) / 1000;
}

function findCommonPhrases(texts: string[], minCount = 5): string[] {
  const ngrams = new Map<string, number>();

  for (const text of texts) {
    const words = text.toLowerCase().split(/\s+/);
    // bigrams and trigrams
    for (let n = 2; n <= 3; n++) {
      for (let i = 0; i <= words.length - n; i++) {
        const gram = words.slice(i, i + n).join(" ");
        ngrams.set(gram, (ngrams.get(gram) ?? 0) + 1);
      }
    }
  }

  return Array.from(ngrams.entries())
    .filter(([, count]) => count >= minCount)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 30)
    .map(([phrase]) => phrase);
}

function findTypicalErrors(texts: string[]): TypicalError[] {
  const errors: TypicalError[] = [];

  // missing apostrophes in common contractions
  const missingApostrophe = [
    { wrong: /\bdont\b/gi, label: "dont (missing apostrophe)" },
    { wrong: /\bcant\b/gi, label: "cant (missing apostrophe)" },
    { wrong: /\bwont\b/gi, label: "wont (missing apostrophe)" },
    { wrong: /\bdidnt\b/gi, label: "didnt (missing apostrophe)" },
    { wrong: /\bisnt\b/gi, label: "isnt (missing apostrophe)" },
    { wrong: /\bthats\b/gi, label: "thats (missing apostrophe)" },
    { wrong: /\btheres\b/gi, label: "theres (missing apostrophe)" },
    { wrong: /\bits\b/gi, label: "its (possessive vs contraction)" },
    { wrong: /\bim\b/gi, label: "im (missing apostrophe)" },
    { wrong: /\byoure\b/gi, label: "youre (missing apostrophe)" },
  ];

  for (const { wrong, label } of missingApostrophe) {
    const matches = texts.filter((t) => wrong.test(t));
    if (matches.length >= 3) {
      errors.push({
        pattern: label,
        frequency: matches.length,
        examples: matches.slice(0, 3).map((t) => t.substring(0, 80)),
      });
    }
  }

  // double spaces
  const doubleSpaces = texts.filter((t) => /  /.test(t));
  if (doubleSpaces.length >= 3) {
    errors.push({
      pattern: "double spaces",
      frequency: doubleSpaces.length,
      examples: doubleSpaces.slice(0, 3).map((t) => t.substring(0, 80)),
    });
  }

  // repeated words
  const repeatedWords = texts.filter((t) =>
    /\b(\w+)\s+\1\b/i.test(t)
  );
  if (repeatedWords.length >= 2) {
    errors.push({
      pattern: "repeated words",
      frequency: repeatedWords.length,
      examples: repeatedWords.slice(0, 3).map((t) => t.substring(0, 80)),
    });
  }

  return errors.sort((a, b) => b.frequency - a.frequency);
}

export function styleProfileToPrompt(profile: StyleProfile): string {
  const lines: string[] = [];

  lines.push("## Writing style characteristics");
  lines.push("");
  lines.push(`- Average message length: ${profile.averageLength} characters`);
  lines.push(
    `- Capitalizes first word: ${pct(profile.capitalizesFirstWord)} of the time`
  );
  lines.push(`- Uses contractions: ${pct(profile.usesContractions)} of the time`);
  lines.push(
    `- Ends with period: ${pct(profile.usesPeriods)} of the time`
  );
  lines.push(`- Uses commas: ${pct(profile.usesCommas)} of the time`);
  lines.push(`- Uses exclamation marks: ${pct(profile.usesExclamation)} of the time`);
  lines.push(`- Uses question marks: ${pct(profile.usesQuestionMark)} of the time`);
  lines.push(`- Uses emoji: ${pct(profile.usesEmoji)} of the time`);
  lines.push(
    `- Short fragment messages: ${pct(profile.sentenceFragmentRatio)} of the time`
  );

  if (profile.lowercaseIRatio > 0.1) {
    lines.push(
      `- Uses lowercase "i" instead of "I": ${pct(profile.lowercaseIRatio)} of the time`
    );
  }

  if (profile.commonPhrases.length > 0) {
    lines.push("");
    lines.push("## Common phrases");
    lines.push(
      profile.commonPhrases.slice(0, 15).map((p) => `- "${p}"`).join("\n")
    );
  }

  if (profile.typicalErrors.length > 0) {
    lines.push("");
    lines.push("## Typical writing quirks/errors");
    for (const err of profile.typicalErrors) {
      lines.push(
        `- ${err.pattern} (${err.frequency} occurrences). Examples: ${err.examples.map((e) => `"${e}"`).join(", ")}`
      );
    }
  }

  return lines.join("\n");
}

function pct(ratio: number): string {
  return `${Math.round(ratio * 100)}%`;
}

File: pkg/style/analyzer.go
package style

import (
	"fmt"
	"math"
	"regexp"
	"sort"
	"strings"

	"github.com/jschell12/replicateme/pkg/corpus"
)

// AnalyzeStyle computes a StyleProfile from a slice of messages.
func AnalyzeStyle(messages []corpus.RawMessage) (corpus.StyleProfile, error) {
	if len(messages) == 0 {
		return corpus.StyleProfile{}, fmt.Errorf("no messages to analyze")
	}

	texts := make([]string, len(messages))
	for i, m := range messages {
		texts[i] = m.Text
	}

	n := float64(len(texts))

	// average length
	totalLen := 0
	for _, t := range texts {
		totalLen += len(t)
	}
	avgLength := float64(totalLen) / n

	// capitalizes first word
	capsFirst := 0
	capsFirstRe := regexp.MustCompile(`^[A-Z]`)
	for _, t := range texts {
		if capsFirstRe.MatchString(t) {
			capsFirst++
		}
	}

	// contractions
	contractionsRe := regexp.MustCompile(`(?i)\b(i'm|i'll|i've|it's|that's|there's|don't|can't|won't|didn't|isn't|aren't|wasn't|weren't|hasn't|haven't|couldn't|wouldn't|shouldn't|y'all|let's|he's|she's|we're|they're|you're|who's|what's|where's|how's)\b`)
	withContractions := 0
	for _, t := range texts {
		if contractionsRe.MatchString(t) {
			withContractions++
		}
	}

	// punctuation
	endsWithPeriod := regexp.MustCompile(`\.\s*$`)
	withPeriods := 0
	withCommas := 0
	withExclamation := 0
	withQuestion := 0
	for _, t := range texts {
		if endsWithPeriod.MatchString(t) {
			withPeriods++
		}
		if strings.Contains(t, ",") {
			withCommas++
		}
		if strings.Contains(t, "!") {
			withExclamation++
		}
		if strings.Contains(t, "?") {
			withQuestion++
		}
	}

	// emoji
	emojiRe := regexp.MustCompile(`[\x{1F600}-\x{1F64F}\x{1F300}-\x{1F5FF}\x{1F680}-\x{1F6FF}\x{1F1E0}-\x{1F1FF}\x{2600}-\x{26FF}\x{2700}-\x{27BF}]`)
	withEmoji := 0
	for _, t := range texts {
		if emojiRe.MatchString(t) {
			withEmoji++
		}
	}

	// sentence fragments: < 30 chars and <= 5 words
	fragmentish := 0
	for _, t := range texts {
		if len(t) < 30 {
			words := strings.Fields(t)
			if len(words) <= 5 {
				fragmentish++
			}
		}
	}

	// lowercase "i" as standalone word (not I'm, I'll, etc.)
	// Go regexp doesn't support lookahead, so match \bi\b then check the next char isn't '
	lowercaseIRe := regexp.MustCompile(`\bi\b`)
	iAnyRe := regexp.MustCompile(`(?i)\b[iI]\b`)
	lowercaseI := 0
	iTotal := 0
	for _, t := range texts {
		if loc := lowercaseIRe.FindStringIndex(t); loc != nil {
			// check char after match isn't apostrophe
			end := loc[1]
			nextChar := string([]rune(t)[end:])
			if len(nextChar) == 0 || (nextChar[0] != '\'') {
				lowercaseI++
			}
		}
		if iAnyRe.MatchString(t) {
			iTotal++
		}
	}

	var lowercaseIRatio float64
	if iTotal > 0 {
		lowercaseIRatio = float64(lowercaseI) / float64(iTotal)
	}

	commonPhrases := findCommonPhrases(texts, 5)
	typicalErrors := findTypicalErrors(texts)

	return corpus.StyleProfile{
		AverageLength:         int(math.Round(avgLength)),
		CapitalizesFirstWord:  round3(float64(capsFirst) / n),
		UsesContractions:      round3(float64(withContractions) / n),
		UsesPeriods:           round3(float64(withPeriods) / n),
		UsesCommas:            round3(float64(withCommas) / n),
		UsesExclamation:       round3(float64(withExclamation) / n),
		UsesQuestionMark:      round3(float64(withQuestion) / n),
		UsesEmoji:             round3(float64(withEmoji) / n),
		CommonPhrases:         commonPhrases,
		TypicalErrors:         typicalErrors,
		SentenceFragmentRatio: round3(float64(fragmentish) / n),
		LowercaseIRatio:       round3(lowercaseIRatio),
	}, nil
}

func round3(v float64) float64 {
	return math.Round(v*1000) / 1000
}

func findCommonPhrases(texts []string, minCount int) []string {
	ngrams := make(map[string]int)

	for _, text := range texts {
		words := strings.Fields(strings.ToLower(text))
		for ngramLen := 2; ngramLen <= 3; ngramLen++ {
			for i := 0; i <= len(words)-ngramLen; i++ {
				gram := strings.Join(words[i:i+ngramLen], " ")
				ngrams[gram]++
			}
		}
	}

	type entry struct {
		phrase string
		count  int
	}
	var entries []entry
	for phrase, count := range ngrams {
		if count >= minCount {
			entries = append(entries, entry{phrase, count})
		}
	}
	sort.Slice(entries, func(i, j int) bool {
		return entries[i].count > entries[j].count
	})

	limit := 30
	if len(entries) < limit {
		limit = len(entries)
	}
	result := make([]string, limit)
	for i := 0; i < limit; i++ {
		result[i] = entries[i].phrase
	}
	return result
}

func findTypicalErrors(texts []string) []corpus.TypicalError {
	var errors []corpus.TypicalError

	// missing apostrophes
	missingApostrophe := []struct {
		re    *regexp.Regexp
		label string
	}{
		{regexp.MustCompile(`(?i)\bdont\b`), "dont (missing apostrophe)"},
		{regexp.MustCompile(`(?i)\bcant\b`), "cant (missing apostrophe)"},
		{regexp.MustCompile(`(?i)\bwont\b`), "wont (missing apostrophe)"},
		{regexp.MustCompile(`(?i)\bdidnt\b`), "didnt (missing apostrophe)"},
		{regexp.MustCompile(`(?i)\bisnt\b`), "isnt (missing apostrophe)"},
		{regexp.MustCompile(`(?i)\bthats\b`), "thats (missing apostrophe)"},
		{regexp.MustCompile(`(?i)\btheres\b`), "theres (missing apostrophe)"},
		{regexp.MustCompile(`(?i)\bits\b`), "its (possessive vs contraction)"},
		{regexp.MustCompile(`(?i)\bim\b`), "im (missing apostrophe)"},
		{regexp.MustCompile(`(?i)\byoure\b`), "youre (missing apostrophe)"},
	}

	for _, ma := range missingApostrophe {
		var matches []string
		for _, t := range texts {
			if ma.re.MatchString(t) {
				matches = append(matches, t)
			}
		}
		if len(matches) >= 3 {
			examples := matches
			if len(examples) > 3 {
				examples = examples[:3]
			}
			trimmed := make([]string, len(examples))
			for i, ex := range examples {
				if len(ex) > 80 {
					trimmed[i] = ex[:80]
				} else {
					trimmed[i] = ex
				}
			}
			errors = append(errors, corpus.TypicalError{
				Pattern:   ma.label,
				Frequency: len(matches),
				Examples:  trimmed,
			})
		}
	}

	// double spaces
	doubleSpaceRe := regexp.MustCompile(`  `)
	var doubleSpaces []string
	for _, t := range texts {
		if doubleSpaceRe.MatchString(t) {
			doubleSpaces = append(doubleSpaces, t)
		}
	}
	if len(doubleSpaces) >= 3 {
		examples := doubleSpaces
		if len(examples) > 3 {
			examples = examples[:3]
		}
		trimmed := make([]string, len(examples))
		for i, ex := range examples {
			if len(ex) > 80 {
				trimmed[i] = ex[:80]
			} else {
				trimmed[i] = ex
			}
		}
		errors = append(errors, corpus.TypicalError{
			Pattern:   "double spaces",
			Frequency: len(doubleSpaces),
			Examples:  trimmed,
		})
	}

	// repeated words (Go RE2 doesn't support backreferences, check manually)
	var repeatedWords []string
	for _, t := range texts {
		words := strings.Fields(strings.ToLower(t))
		for i := 0; i+1 < len(words); i++ {
			if words[i] == words[i+1] && len(words[i]) > 1 {
				repeatedWords = append(repeatedWords, t)
				break
			}
		}
	}
	if len(repeatedWords) >= 2 {
		examples := repeatedWords
		if len(examples) > 3 {
			examples = examples[:3]
		}
		trimmed := make([]string, len(examples))
		for i, ex := range examples {
			if len(ex) > 80 {
				trimmed[i] = ex[:80]
			} else {
				trimmed[i] = ex
			}
		}
		errors = append(errors, corpus.TypicalError{
			Pattern:   "repeated words",
			Frequency: len(repeatedWords),
			Examples:  trimmed,
		})
	}

	// sort by frequency descending
	sort.Slice(errors, func(i, j int) bool {
		return errors[i].Frequency > errors[j].Frequency
	})

	return errors
}

// StyleProfileToPrompt renders a profile as a human-readable prompt section.
func StyleProfileToPrompt(profile corpus.StyleProfile) string {
	var sb strings.Builder

	sb.WriteString("## Writing style characteristics\n\n")
	sb.WriteString(fmt.Sprintf("- Average message length: %d characters\n", profile.AverageLength))
	sb.WriteString(fmt.Sprintf("- Capitalizes first word: %s of the time\n", pct(profile.CapitalizesFirstWord)))
	sb.WriteString(fmt.Sprintf("- Uses contractions: %s of the time\n", pct(profile.UsesContractions)))
	sb.WriteString(fmt.Sprintf("- Ends with period: %s of the time\n", pct(profile.UsesPeriods)))
	sb.WriteString(fmt.Sprintf("- Uses commas: %s of the time\n", pct(profile.UsesCommas)))
	sb.WriteString(fmt.Sprintf("- Uses exclamation marks: %s of the time\n", pct(profile.UsesExclamation)))
	sb.WriteString(fmt.Sprintf("- Uses question marks: %s of the time\n", pct(profile.UsesQuestionMark)))
	sb.WriteString(fmt.Sprintf("- Uses emoji: %s of the time\n", pct(profile.UsesEmoji)))
	sb.WriteString(fmt.Sprintf("- Short fragment messages: %s of the time\n", pct(profile.SentenceFragmentRatio)))

	if profile.LowercaseIRatio > 0.1 {
		sb.WriteString(fmt.Sprintf("- Uses lowercase \"i\" instead of \"I\": %s of the time\n", pct(profile.LowercaseIRatio)))
	}

	if len(profile.CommonPhrases) > 0 {
		sb.WriteString("\n## Common phrases\n")
		limit := 15
		if len(profile.CommonPhrases) < limit {
			limit = len(profile.CommonPhrases)
		}
		for _, p := range profile.CommonPhrases[:limit] {
			sb.WriteString(fmt.Sprintf("- \"%s\"\n", p))
		}
	}

	if len(profile.TypicalErrors) > 0 {
		sb.WriteString("\n## Typical writing quirks/errors\n")
		for _, err := range profile.TypicalErrors {
			quoted := make([]string, len(err.Examples))
			for i, ex := range err.Examples {
				quoted[i] = fmt.Sprintf("\"%s\"", ex)
			}
			sb.WriteString(fmt.Sprintf("- %s (%d occurrences). Examples: %s\n",
				err.Pattern, err.Frequency, strings.Join(quoted, ", ")))
		}
	}

	return sb.String()
}

func pct(ratio float64) string {
	return fmt.Sprintf("%d%%", int(math.Round(ratio*100)))
}
